hack/generator/pkg/codegen/pipeline_determine_resource_ownership.go

```diff
@@ -0,0 +1,293 @@
+	if isObject && isObjectOneOfObject(resourcesPropertyTypeAsObject) {
+		return resolveResourcesTypeNames(resourcesPropertyTypeDef.Name(), resourcesPropertyTypeAsObject)
+	} else {
+		return []astmodel.TypeName{resourcesPropertyTypeDef.Name()}, nil
+	}
+}
+
+func isObjectOneOfObject(o *astmodel.ObjectType) bool {
+	return o.HasFunctionWithName(astmodel.JSONMarshalFunctionName)
+}
+
+func updateChildResourceDefinitionsWithOwner(
+	definitions astmodel.Types,
+	childResourceTypeNames []astmodel.TypeName,
+	owningResourceName astmodel.TypeName,
+	updatedDefs astmodel.Types) error {
+
+	for _, typeName := range childResourceTypeNames {
+		// If the typename ends in ChildResource, remove that
+		if strings.HasSuffix(typeName.Name(), "ChildResource") {
+			typeName = astmodel.MakeTypeName(typeName.PackageReference, strings.TrimSuffix(typeName.Name(), "ChildResource"))
+		}
+
+		// TODO: These are types that cause us trouble... a lot of them use allof inheritance.
+		// TODO: I think for these we will need to walk the graph of types and do a structural
+		// TODO: equality check to find the name of the actual resource, but we can't do that check
+		// TODO: now because these types allOf inherit from resourceBase and the actual resources
+		// TODO: being referenced do not. See:
+		if typeName.Name() == "VirtualMachinesSpec_Resources" || // Uses allof inheritance
+			typeName.Name() == "AccountSpec_Resources" || // Uses allof inheritance
+			typeName.Name() == "SitesSpec_Resources" || // Uses allof inheritance
+			typeName.Name() == "NamespacesSpec_Resources" || // Uses allof inheritance
+			typeName.Name() == "VaultsSpec_Resources" || // Uses allof inheritance
+			// Bug in spec which there is a PR out for: https://github.com/Azure/azure-resource-manager-schemas/pull/1071
+			// TODO: remove the below once PR is merged
+			typeName.Name() == "ServersAdministrators" ||
+			typeName.Name() == "ExtensionsChild" {
+			continue
+		}
+
+		// Confirm the type really exists
+		childResourceDef, ok := definitions[typeName]
+		if !ok {
+			return errors.Errorf("couldn't find child resource type %s", typeName)
+		}
+
+		// Update the definition of the child resource type to point to its owner
+		childResource, ok := childResourceDef.Type().(*astmodel.ResourceType)
+		if !ok {
+			return errors.Errorf("child resource %s not of type *astmodel.ResourceType, instead %T", typeName, childResourceDef.Type())
+		}
+
+		childResourceDef = childResourceDef.WithType(childResource.WithOwner(&owningResourceName))
+
+		// There can be overwrites here because updatedDefs contains all resources we're processing and it's possible
+		// to have a resource graph where A owns B owns C, but we process B first, resulting in B having the "Resources" property
+		// removed and being put into the updatedDefs, and then we process A which will try to add an owner to B which is already
+		// in updatedDefs. Because order doesn't matter here we're okay with an overwrite.
```

Comment:
`B'` will be updated. This is what the check on line `242` is about:
```
               // Check if we've already modified this type
		if updatedDef, ok := updatedDefs[def.Name()]; ok {
			def = updatedDef
		}
```

I've added a _little_ clarification to the comment on like 242 - let me know if there's something else I can say here to make it clearer.

Reply:
_Really_ good catch. In another place in the code where we do this, I do the following:

```
		// Check if we've already modified this type - we need to use the already modified value
		if updatedDef, ok := updatedDefs[def.Name()]; ok {
			def = updatedDef
		}
```

I didn't do that here -- but we 100% should.

I'll add a test for this and fix up the code to get the value from `updateDefs` if it's there, so that we're using `B'` when making the subsequent modification

Reply:
Hah, okay so actually, this comment is super misleading (it confused me too, and I wrote it!) It's actually entirely wrong. The problem is that it is conflating `Spec`'s and `Resource`s.

`Owner` is on the `Resource`, but `Resources` is on the `Spec`, so there is no chance of conflict, and I can actually delete this comment and use the built in `Add`.

Reply:
I did modify a test so that it would catch this in case it was possible though (and then it wasn't reproing, which is what made me re-examine the comment and realize that actually there is no chance of collision!)